### PR TITLE
OECD Regional Demography Import (Integration Test coming soon)

### PR DESCRIPTION
**Pull request by wh1210**
This PR contains 5 datasets import for OECD:
- deaths (Deaths by 5-year age groups,small regions TL3)
- life_expectancy_and_mortality (Life Expectancy and Mortality, large TL2 and small TL3 regions)
- pop_density (Population density and area, large TL2 and small TL3 regions)
- population_tl2 (Population by 5-year age groups, large regions TL2)
- population_tl3 (Population by 5-year age groups, small regions TL3)
Also improve the accuracy for geoId issue strategy(P0: country P1: Nuts P2: resolver) and fix some typos/errors.

### CODE REVIEW

File: scripts/oecd/regional_demography/life_expectancy_and_mortality/config.txt [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import: {
+    import_name: "OECD_life_expectancy_and_mortality"
+    provenance_url: "https://stats.oecd.org"
+    provenance_description: "Various regional statistics measured by OECD(Organisation for Economic Co-operation and Development)."
```
[the remark this answers is not in the document]
Comment (wh1210) [reply scripts/oecd/regional_demography/life_expectancy_and_mortality/config.txt:4]: done

File: scripts/oecd/regional_demography/life_expectancy_and_mortality/OECD_life_expectancy_and_mortality_stat_vars.mcf [hunk 2]
```diff
@@ -38,7 +38,7 @@ measuredProperty: dcs:count
 statType: dcs:measuredValue
 measurementDenominator: dcs:Count_Person
```
[the remark this answers is not in the document]
Comment (wh1210) [reply scripts/oecd/regional_demography/life_expectancy_and_mortality/OECD_life_expectancy_and_mortality_stat_vars.mcf]: Done

File: scripts/oecd/regional_demography/README.md [hunk 3]
```diff
@@ -20,3 +20,40 @@ go run ../../../tools/place_name_resolver/resolver.go --in_csv_path=geos_cleaned
...
+All the MCF files are named by `OECD_{subdirectory name}_stat_vars.mcf`. Except population TL2 and TL3, these two datasets have the same MCF files.
+
+### Template MCFs
+All the template MCF files are named by `OECD_{subdirectory name}.tmcf`.
+
+### Script
+The script file `preprocess_csv.py` are used to generate cleaned csv files and template MCFs.
+
+## Generating Artifacts
+To generate the cleaned csv and template MCF file, run
```
[the remark this answers is not in the document]
Comment (wh1210) [reply scripts/oecd/regional_demography/README.md:56]: Done

File: scripts/oecd/regional_demography/README.md [hunk 4]
```diff
@@ -20,3 +20,40 @@ go run ../../../tools/place_name_resolver/resolver.go --in_csv_path=geos_cleaned
...
+All the cleaned data are named by `OECD_{subdirectory name}_cleaned.csv`.
+
+### StatisticalVariables(MCF) files
+All the MCF files are named by `OECD_{subdirectory name}_stat_vars.mcf`. Except population TL2 and TL3, these two datasets have the same MCF files.
+
+### Template MCFs
+All the template MCF files are named by `OECD_{subdirectory name}.tmcf`.
+
+### Script
+The script file `preprocess_csv.py` are used to generate cleaned csv files and template MCFs.
```
[the remark this answers is not in the document]
Comment (wh1210) [reply scripts/oecd/regional_demography/README.md:53]: Done

File: scripts/oecd/regional_demography/population_tl2/preprocess_csv.py [hunk 5]
```diff
@@ -12,32 +12,28 @@
...
+df = df[['TL', 'REG_ID', 'Region', 'VAR', 'SEX', 'Year', 'Value']]
 # First remove geos with names that we don't have mappings to dcid for.
 name2dcid = dict(json.loads(open('../name2dcid.json').read()))
-df = df[df['Region'].isin(name2dcid.keys())]
+nuts = dict(json.loads(open('../region_nuts_codes.json').read()))
+df = df[df['REG_ID'].isin(nuts.keys()) | df['Region'].isin(name2dcid.keys())]
 # Second, replace the names with dcids
-df.replace({'Region': name2dcid}, inplace=True)
+df['Region'] = df.apply(lambda row: generate_geo_id(row, nuts, name2dcid), axis=1)
+# df.replace({'Region': name2dcid}, inplace=True)
```
[the remark this answers is not in the document]
Comment (wh1210) [reply scripts/oecd/regional_demography/population_tl2/preprocess_csv.py:32]: Done

File: scripts/oecd/regional_demography/utils.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        columns.append(''.join(column))
+
+    df.columns = columns
+    return df.reset_index()
+
+
+# Generate the dcid info.
+def generate_geo_id(row, nuts, name2dcid):
+    if str(row['TL']) == '1':
+        # print("TL1: " + "Country/" + row['REG_ID'])
```
[the remark this answers is not in the document]
Comment (wh1210) [reply scripts/oecd/regional_demography/utils.py:31]: Done

File: scripts/oecd/regional_demography/utils.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+        return "Country/" + row['REG_ID']
```
[the remark this answers is not in the document]
Comment (wh1210) [reply scripts/oecd/regional_demography/utils.py]: Done

File: scripts/oecd/regional_demography/utils.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    elif row['REG_ID'] in nuts.keys():
+        # print("NUTS: " + "dcid:nuts/" + row['REG_ID'])
+        return "dcid:nuts/" + row['REG_ID']
```
[the remark this answers is not in the document]
Comment (wh1210) [reply scripts/oecd/regional_demography/utils.py:35]: Yes, I leave "dcid" before for all.

File: scripts/oecd/regional_demography/region_nuts_codes.json [hunk 9]
```diff
@@ -0,0 +1 @@
...7ankiri", "TR823": "Sinop", "TR83": "Samsun, Tokat, \u00c7orum, Amasya", "TR831": "Samsun", "TR832": "Tokat", "TR833": "\u00c7orum", "TR834": "Amasya", "TR9": "Dogu Karadeniz", "TR90": "Trabzon, Ordu, Giresun, Rize, Artvin, G\u00fcm\u00fcshane", "TR901": "Trabzon", "TR902": "Ordu", "TR903": "Giresun", "TR904": "Rize", "TR905": "Artvin", "TR906": "G\u00fcm\u00fcshane", "TRA": "Kuzeydogu Anadolu", "TRA1": "Erzurum, Erzincan, Bayburt", "TRA11": "Erzurum", "TRA12": "Erzincan", "TRA13": "Bayburt", "TRA2": "Agri, Kars, Igdir, Ardahan", "TRA21": "Agri", "TRA22": "Kars", "TRA23": "Igdir", "TRA24": "Ardahan", "TRB": "Ortadogu Anadolu", "TRB1": "Malatya, Elazig, Bing\u00f6l, Tunceli", "TRB11": "Malatya", "TRB12": "Elazig", "TRB13": "Bing\u00f6l", "TRB14": "Tunceli", "TRB2": "Van, Mus, Bitlis, Hakkari", "TRB21": "Van", "TRB22": "Mus", "TRB23": "Bitlis", "TRB24": "Hakkari", "TRC": "G\u00fcneydogu Anadolu", "TRC1": "Gaziantep, Adiyaman, Kilis", "TRC11": "Gaziantep", "TRC12": "Adiyaman", "TRC13": "Kilis", "TRC2": "Sanliurfa, Diyarbakir", "TRC21": "Sanliurfa", "TRC22": "Diyarbakir", "TRC3": "Mardin, Batman, Sirnak, Siirt", "TRC31": "Mardin", "TRC32": "Batman", "TRC33": "Sirnak", "TRC34": "Siirt"}
```
[the remark this answers is not in the document]
Comment (wh1210) [reply scripts/oecd/regional_demography/region_nuts_codes.json:1]: Done

File: scripts/oecd/regional_demography/population_tl2/config.txt [hunk 10]
```diff
@@ -0,0 +1,8 @@
+import: {
+    import_name: "OECD_population_tl2"
```
[the remark this answers is not in the document]
Comment (wh1210) [reply scripts/oecd/regional_demography/population_tl2/config.txt:2]: Is that mean we only need one config file in regional_demography directory?

File: scripts/oecd/regional_demography/deaths/preprocess_csv.py [hunk 11]
```diff
@@ -12,29 +12,23 @@
...
 df = pd.read_csv('REGION_DEMOGR_death_tl3.csv')
-
+df = df[['TL', 'REG_ID', 'Region', 'VAR', 'SEX', 'Year', 'Value']]
 # First remove geos with names that we don't have mappings to dcid for.
 name2dcid = dict(json.loads(open('../name2dcid.json').read()))
-df = df[df['Region'].isin(name2dcid.keys())]
-# Second, replace the names with dcids.
-df.replace({'Region': name2dcid}, inplace=True)
+nuts = dict(json.loads(open('../region_nuts_codes.json').read()))
+df = df[df['REG_ID'].isin(nuts.keys()) | df['Region'].isin(name2dcid.keys())]
```
[the remark this answers is not in the document]
Comment (wh1210) [reply scripts/oecd/regional_demography/deaths/preprocess_csv.py:29]: Yep, I've checked that. there are very few countries can be dropped. I checked with pop_density dataset. ('IRL', 'OECD', 'RUS') are the only three one in TL1 being dropped.

[the remark this answers is not in the document]
Comment (tjann) [reply scripts/oecd/regional_demography/deaths/preprocess_csv.py:29]: So IRL and RUS are not in the final result?

File: scripts/oecd/regional_demography/utils.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (tjann) [reply scripts/oecd/regional_demography/utils.py:33]: Thanks for the change!

File: scripts/oecd/regional_demography/deaths/preprocess_csv.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (wh1210) [reply scripts/oecd/regional_demography/deaths/preprocess_csv.py:29]: Yep

File: scripts/oecd/regional_demography/population_tl2/config.txt (hunk 10 above)
[the remark this answers is not in the document]
Comment (tjann) [reply scripts/oecd/regional_demography/population_tl2/config.txt:2]: Yes

File: scripts/oecd/regional_demography/deaths/preprocess_csv.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (tjann) [reply scripts/oecd/regional_demography/deaths/preprocess_csv.py:29]: Can you file a bug and fix that later?

[the remark this answers is not in the document]
Comment (wh1210) [reply scripts/oecd/regional_demography/deaths/preprocess_csv.py:29]: Yes.

**Comment by wh1210** [comment]
> Can you add some testing for your code?

Added~ Also dup issue fixed.

File: scripts/oecd/regional_demography/population_tl2/config.txt (hunk 10 above)
[the remark this answers is not in the document]
Comment (wh1210) [reply scripts/oecd/regional_demography/population_tl2/config.txt]: Done

**Comment by tjann** [comment]
I will approve and merge this PR, but will file a bug for integration test.

**Pull request merged**